### PR TITLE
Document functions generated in endpoint as callbacks

### PR DESCRIPTION
**Pull request by michalmuskala**
This makes it easy to search for those functions using the search box in
hexdocs.

A similar pattern is used for documenting Ecto.Repo.

### CODE REVIEW

File: lib/phoenix/endpoint.ex [hunk 1]
```diff
@@ -365,16 +352,123 @@ defmodule Phoenix.Endpoint do
...
   `Phoenix.Endpoint.instrument/4` macro. Refer to its documentation for more
   information.
 
   """
 
+  @type topic :: String.t
+  @type event :: String.t
+  @type msg :: map
+
+  # Configuration
```
Comment (josevalim) [inline lib/phoenix/endpoint.ex]: Please add a new line.

File: lib/phoenix/endpoint.ex [hunk 2]
```diff
@@ -365,16 +352,123 @@ defmodule Phoenix.Endpoint do
...
+  @callback broadcast_from(from :: pid, topic, event, msg) :: :ok | {:error, term}
+
+  @doc """
+  Broadcasts a `msg` from the given `from` as `event` in the given `topic`.
+
+  Raises in case of failures.
+  """
+  @callback broadcast_from!(from :: pid, topic, event, msg) :: :ok | no_return
+
+  # Instrumentation
```
Comment (josevalim) [inline lib/phoenix/endpoint.ex]: New line here.

File: lib/phoenix/endpoint.ex [hunk 3]
```diff
@@ -365,16 +352,123 @@ defmodule Phoenix.Endpoint do
...
+  @doc """
+  Allows instrumenting operation defined by `function`.
+
+  `runtime_metadata` may be omitted and defaults to `nil`.
+
+  Read more about instrumentation in the "Instrumentation" section.
+  """
+  @macrocallback instrument(instrument_event :: Macro.t, runtime_metadata :: Macro.t, funcion :: Macro.t) :: Macro.t
+
+
```
Comment (josevalim) [inline lib/phoenix/endpoint.ex]: ✂️

**Comment by josevalim** [comment]
:+1:

File: lib/phoenix/endpoint.ex [hunk 4]
```diff
@@ -193,60 +193,47 @@ defmodule Phoenix.Endpoint do
...
   functions that are automatically defined in your endpoint.
 
   #### Paths and URLs
 
-    * `struct_url()` - generates the endpoint base URL, but as a `URI` struct
-    * `url()` - generates the endpoint base URL without any path information
-    * `path(path)` - generates the path information when routing to this endpoint
-    * `static_url()` - generates the static URL without any path information
-    * `static_path(path)` - generates a route to a static file in `priv/static`
+    * `c:struct_url/0` and `c:url/0` - that generate the endpoint base URL
```
Comment (whatyouhide) [inline lib/phoenix/endpoint.ex]: Why merge these two? And in any case, there's a superfluous "that" after the `-` :)

**Comment by josevalim** [comment]
I agree with @whatyouhide that we can probably get rid of the module summaries.

**Comment by michalmuskala** [comment]
I tried limiting the overall changes as much as possible. The module docs are designed around a logical sequence, so I wanted to keep this. It also allows grouping the callbacks, which is not possible with "just" docs. I'll try to maybe only list the functions and leave out any descriptions.

**Comment by whatyouhide** [comment]
@michalmuskala I meant to keep the sections idea, but ditch the small summary of each function in the moduledoc. In the moduledoc you could say: "callbacks related to endpoint are x, y, and z", "callbacks related to ... are ..." and so on.

**Comment by michalmuskala** [comment]
I updated it to only list the functions grouped by category in the moduledocs.

**Comment by chrismccord** [comment]
❤️❤️❤️🐥🔥

**Pull request merged**